File: AzimMuradov/Term2/Block3/Task7/lib/meteo-storm-glass/src/main/kotlin/meteo/stormglass/data/StormGlassApi.kt
package meteo.stormglass.data

import meteo.data.MeteoApi
import meteo.domain.entity.Location
import java.net.URI
import java.net.http.HttpRequest
import java.time.Instant
import java.time.temporal.ChronoUnit

object StormGlassApi : MeteoApi {

    override fun createGetWeatherRequest(location: Location, key: String): HttpRequest =
        HttpRequest.newBuilder().apply {
            val time = currentTimeInEpochSeconds()
            val (lat, lon) = location
            val query = listOf(
                "lat=$lat", "lng=$lon",
                "params=${params.joinToString(separator = ",")}",
                "start=$time", "end=$time",
                "source=$SG"
            ).joinToString(separator = "&")

            uri(URI.create("$API_ENDPOINT?$query"))
            GET()
            header("Authorization", key)
        }.build()


    private const val API_ENDPOINT = "https://api.stormglass.io/v2/weather/point"

    private val params = listOf(
        "airTemperature",
        "cloudCover",
        "humidity",
        "precipitation",
        "windDirection",
        "windSpeed"
    )

    internal const val SG = "sg"

    private fun currentTimeInEpochSeconds() = Instant.now().truncatedTo(ChronoUnit.HOURS).epochSecond
}

File: Daniil-Kadochnikov/Second Term/ThirdBlock/Weather/WeatherAPI/StormGlassAPI.cs
using Newtonsoft.Json.Linq;
using System.Net.Http.Headers;
using WeatherForecastModel;

namespace WeatherAPI
{
	public class StormGlassAPI : AWeatherAPI
	{
		public StormGlassAPI() : base("https://api.stormglass.io/v2/weather/point?lat=59.93863&lng=30.31413&params=airTemperature,cloudCover,humidity,precipitation,windDirection,windSpeed", "StormGlass.io") { }

		public override async Task<string> GetDataAsync()
		{
			try
			{
				Client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(Keys.StormGlassIoAPIkey);
				HttpResponseMessage responseHttp = await Client.GetAsync(URL + $"&start={new DateTimeOffset(DateTime.UtcNow).ToUnixTimeSeconds()}&end={new DateTimeOffset(DateTime.UtcNow).ToUnixTimeSeconds()}&source=sg"); // + $"&start={DateTime.Now}&end={DateTime.Now}&source=sg"
				responseHttp.EnsureSuccessStatusCode();

				return await responseHttp.Content.ReadAsStringAsync();
			}
			catch (Exception e)
			{
				Flag = false;
				throw e;
			}
		}

		public override WeatherModel GetWeatherModelAsync(string responseString)
		{
			JObject json = JObject.Parse(responseString);
			JObject? values = (JObject?)json["hours"][0];
			WeatherModel? model = new WeatherModel();
			model.CloudCover = (string)values["cloudCover"]["sg"];
			model.Humidity = (string)values["humidity"]["sg"];
			model.PrecipitationProbability = (string)values["precipitation"]["sg"];
			model.Temperature = (string)values["airTemperature"]["sg"];
			model.WindDirection = (string)values["windDirection"]["sg"];
			model.WindSpeed = (string)values["windSpeed"]["sg"];
			return model;
		}
	}
}